solutions/cses-2101.mdx

```diff
@@ -0,0 +1,166 @@
+---
+id: cses-2101
+source: CSES
+title: New Roads Queries
+author: Dustin Miao
+---
+
+Let's construct a graph where edges have weights equal to the time they are added. The problem boils down to efficiently finding the path with the minimum maximum edge for each query. 
+
+Let's consider a single query between nodes $u, v$. One way of approaching this is to sort the edge weights from smallest to largest, and then add edges one by one until $u$ and $v$ are connected. If we use a [disjoint set union](/gold/dsu), the time complexity for each query is $\mathcal{O}(m \cdot \alpha(n))$, which is too slow.
+
+However, this hints at a faster solution. Note that our slow approach bears many similarities with [Kruskal's MST algorithm](/gold/mst/#kruskals), which implies that the path between $u$ and $v$ on any minimum spanning tree will minimize the maximum edge weight. For a more formal proof, see [here](https://www.geeksforgeeks.org/minimum-bottleneck-spanning-treembst/). 
+****
+All that is left to do is to efficiently query for the minimum node on a tree path. One way to do this is with [binary jumping](/plat/binary-jump). 
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(m \cdot \alpha(n) + n\log n)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+ 
+/** Disjoint Set Union with Small-to-Large Merging 
+ *  Verified: https://judge.yosupo.jp/problem/unionfind */ 
+template<size_t N> struct UnionFind {
+	int par[N], sze[N], max_size;
+ 
+	UnionFind(int n = N) { init(n); }
+ 
+	/** Complexity: O(n) */ 
+	void init(int n = N) {
+		iota(par, par + n, 0);
+		fill(sze, sze + n, 1);
+		max_size = 1;
+	}
+ 
+	/** Complexity: O(alpha(n)) */ 
+	int find(int a) {
+		if (a == par[a])
+			return a;
+		return par[a] = find(par[a]);
+	}
+ 
+	/** Complexity: O(alpha(n)) */ 
+	bool merge(int a, int b) {
+		a = find(a), b = find(b);
+		if (a == b) 
+			return 0;
+		
+		if (sze[a] > sze[b]) {
+			par[b] = a;
+			sze[a] += sze[b];
+			max_size = max(max_size, sze[a]);
+		}
+		else {
+			par[a] = b;
+			sze[b] += sze[a];
+			max_size = max(max_size, sze[b]);
+		}
+		return 1;
+	}
+ 
+	/** Complexity: O(alpha(n)) */ 
+	int size(int a) { 
+		return sze[find(a)]; 
+	}
+};
+ 
+int N, M, Q;
+vector<pair<int, int>> G[200005];
+UnionFind<200005> dsu;
```

Comment:
so magic numbers are bad blah blah blah
put 200k in a number called `MAX_N` or something and then in declaring these arrays
do something like `int arr[MAX_N + 1]`